services/venice-controller/src/main/java/com/linkedin/venice/controller/systemstore/SystemStoreHealthCheckService.java

```diff
@@ -0,0 +1,179 @@
+import com.linkedin.venice.common.VeniceSystemStoreUtils;
+import com.linkedin.venice.exceptions.VeniceException;
+import com.linkedin.venice.meta.ReadWriteStoreRepository;
+import com.linkedin.venice.meta.Store;
+import com.linkedin.venice.pushstatushelper.PushStatusStoreReader;
+import com.linkedin.venice.pushstatushelper.PushStatusStoreWriter;
+import com.linkedin.venice.service.AbstractVeniceService;
+import com.linkedin.venice.system.store.MetaStoreReader;
+import com.linkedin.venice.system.store.MetaStoreWriter;
+import com.linkedin.venice.utils.RetryUtils;
+import com.linkedin.venice.utils.concurrent.VeniceConcurrentHashMap;
+import java.time.Duration;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+
+/**
+ * This class is used to schedule periodic check of user system stores in Venice cluster to make sure system stores has
+ * good version and is ingesting correctly. This service is expected to be setup in the leader controller of the cluster,
+ * and it will periodically scan all user system stores and collect bad system stores. Parent controller will hit the
+ * controller endpoint to collect information about bad system stores.
+ */
+public class SystemStoreHealthCheckService extends AbstractVeniceService {
+  public static final Logger LOGGER = LogManager.getLogger(SystemStoreHealthCheckService.class);
+  private final ReadWriteStoreRepository storeRepository;
+  private final PushStatusStoreReader pushStatusStoreReader;
+  private final PushStatusStoreWriter pushStatusStoreWriter;
+  private final MetaStoreReader metaStoreReader;
+  private final MetaStoreWriter metaStoreWriter;
+
+  private final int checkPeriodInSeconds;
+  private final AtomicBoolean isRunning = new AtomicBoolean(false);
+  private final AtomicReference<Set<String>> unhealthySystemStoreSet = new AtomicReference<>();
+
+  private ScheduledExecutorService checkServiceExecutor;
+
+  public SystemStoreHealthCheckService(
+      ReadWriteStoreRepository storeRepository,
+      MetaStoreReader metaStoreReader,
+      MetaStoreWriter metaStoreWriter,
+      PushStatusStoreReader pushStatusStoreReader,
+      PushStatusStoreWriter pushStatusStoreWriter,
+      int systemStoreCheckPeriodInSeconds) {
+    this.storeRepository = storeRepository;
+    this.metaStoreWriter = metaStoreWriter;
+    this.metaStoreReader = metaStoreReader;
+    this.pushStatusStoreWriter = pushStatusStoreWriter;
+    this.pushStatusStoreReader = pushStatusStoreReader;
+    this.checkPeriodInSeconds = systemStoreCheckPeriodInSeconds;
+    this.unhealthySystemStoreSet.set(new HashSet<>());
```

Comment:
We don't really need to new the HashSet from memory, do we? We just periodically update `unhealthySystemStoreSet` to point to the most recent data.

Reply:
It is either here, or in the get method. If the parent query comes first before first iteration, you will need to have one to return.

Reply:
Hmm... why can't return a null pointer if parent query comes before the 1st iteration?

Reply:
Nullpointer in my opinion is dangerous, we will have to handle null pointer anyway, some time some where. I prefer we do proper initialization in child controller to avoid this handling. I think an alternative way is to check in get method, but  I think the effect is the same.

Reply:
Not quite sure about the exact concern, but maybe we can live with it. It might have some performance trade-offs though, e.g. memory allocation/deallocation on the heap which itself is not cheap and extra call to constructor.

I think the change looks okay to me, please address other reviews, thanks Jialin!